lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,526 @@
+        /// Removes the task from running tasks if the task was running
+        /// Updates the task state to fail based on the error message in the failed task
+        /// </summary>
+        /// <param name="failedTask"></param>
+        internal void SetFailedRunningTask(IFailedTask failedTask)
+        {
+            RemoveRunningTask(failedTask.Id);
+            UpdateState(failedTask.Id, GetTaskErrorEvent(failedTask));
+        }
+
+        /// <summary>
+        /// This method is called when receiving IFailedTask event during system shutting down. 
+        /// If the task failed because it receives the close command from driver, update the task state to TaskClosedByDriver.
+        /// Task could fail by communication error or any other application or system error during this time, as long as it is not 
+        /// TaskFailedByEvaluatorFailure, update the task state based on the error received. 
+        /// </summary>
+        /// <param name="failedTask"></param>
+        internal void SetFailedTaskInShuttingDown(IFailedTask failedTask)
+        {
+            if (TaskState(failedTask.Id) == StateMachine.TaskState.TaskWaitingForClose)
+            {
+                UpdateState(failedTask.Id, TaskStateEvent.ClosedTask);
+            }
+            else if (TaskState(failedTask.Id) != StateMachine.TaskState.TaskFailedByEvaluatorFailure)
+            {
+                UpdateState(failedTask.Id, GetTaskErrorEvent(failedTask));
+            }
+        }
+
+        /// <summary>
+        /// This method is called when receiving an IFailedEvaluator event during TaskSubmitted, TaskRunning or system ShuttingDown.
+        /// Removes the task from RunningTasks if the task associated with the FailedEvaluator is present and running. 
+        /// Sets the task state to TaskFailedByEvaluatorFailure 
+        /// </summary>
+        /// <param name="failedEvaluator"></param>
+        internal void SetTaskFailByEvaluator(IFailedEvaluator failedEvaluator)
+        {
+            if (failedEvaluator.FailedTask.IsPresent())
+            {
+                var taskId = failedEvaluator.FailedTask.Value.Id;
+                if (TaskState(taskId) == StateMachine.TaskState.TaskRunning)
+                {
+                    RemoveRunningTask(taskId);
+                }
+
+                if (TaskState(taskId) == StateMachine.TaskState.TaskWaitingForClose)
+                {
+                    UpdateState(taskId, TaskStateEvent.ClosedTask);
+                }
+                else
+                {
+                    UpdateState(taskId, TaskStateEvent.FailedTaskEvaluatorError);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Removes a task from running tasks if it exists in the running tasks collection
+        /// </summary>
+        /// <param name="taskId"></param>
+        private void RemoveRunningTask(string taskId)
+        {
+            if (_runningTasks.ContainsKey(taskId))
```

Comment:
It can be called in a scenario where the task may not run yet, so there is no exception if it doesn't exist. 

To make it more strict, I will enforce the contains checking in this method. In the case where the task may not run, I will call  _runningTasks.Remove(taskId) directly regardless what bool it would return.